File: 3349-maximum-length-substring-with-two-occurrences/maximum-length-substring-with-two-occurrences.cpp
class Solution {
public:
    int maximumLengthSubstring(string s) {
        int n = s.size();
        int l, r = 0;
        unordered_map<char, int> hash;
        int maxlen = 0;
        while (r < n) {
             hash[s[r]]++;
            while (hash[s[r]] > 2) {
                hash[s[l]]--;
                l++;
            }
            maxlen = max(maxlen, r - l + 1);
           
            r++;
        }
        return maxlen;
    }
};

File: 3349-maximum-length-substring-with-two-occurrences/maximum-length-substring-with-two-occurrences.py
class Solution:
    def maximumLengthSubstring(self, s: str) -> int:
        n = len(s)
        l = 0
        r = 0
        counts = {}
        max_len = 0
        

        while r < n:
            counts[s[r]] = counts.get(s[r], 0) + 1
            
            
            while counts[s[r]] > 2:
                counts[s[l]] -= 1
                l += 1
                
            max_len = max(max_len, r - l + 1)
            r += 1
            
        return max_len
        